plotter/histogram_gnuplotstyle_test.go

```diff
@@ -0,0 +1,123 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package plotter
+
+import (
+	"image/color"
+	"log"
+	"math"
+	"math/rand"
+	"testing"
+
+	"github.com/gonum/plot"
+	"github.com/gonum/plot/vg"
+	"github.com/gonum/plot/vg/draw"
+)
+
+// An example of making a histogram with a custom style.
+func ExampleHistogramGnuplot() {
+	// stdNorm returns the probability of drawing a
+	// value from a standard normal distribution.
+	stdNorm := func(x float64) float64 {
+		const sigma = 1.0
+		const mu = 0.0
+		const root2π = 2.50662827459517818309
+		return 1.0 / (sigma * root2π) * math.Exp(-((x-mu)*(x-mu))/(2*sigma*sigma))
+	}
+
+	n := 10000
+	vals := make(Values, n)
+	for i := 0; i < n; i++ {
+		vals[i] = rand.NormFloat64()
+	}
+
+	p, err := plot.New()
+	if err != nil {
+		log.Panic(err)
+	}
+	p.Title.Text = "Histogram"
+	p.X.Label.Text = "X-axis"
+	p.Y.Label.Text = "Y-axis"
+	p.Style = GnuplotStyle{}
+
+	h, err := NewHist(vals, 16)
+	if err != nil {
+		log.Panic(err)
+	}
+	h.Normalize(1)
+	p.Add(h)
+
+	// The normal distribution function
+	norm := NewFunction(stdNorm)
+	norm.Color = color.RGBA{R: 255, A: 255}
+	norm.Width = vg.Points(2)
+	p.Add(norm)
+
+	err = p.Save(200, 200, "testdata/histogram-gnuplot-style.png")
+	if err != nil {
+		log.Panic(err)
+	}
+}
+
+func TestHistogramGnuplot(t *testing.T) {
+	checkPlot(ExampleHistogramGnuplot, t, "histogram-gnuplot-style.png")
+}
+
+type GnuplotStyle struct{}
+
+func (s GnuplotStyle) DrawPlot(p *plot.Plot, c draw.Canvas) {
+	if p.BackgroundColor != nil {
+		c.SetColor(p.BackgroundColor)
+		c.Fill(c.Rectangle.Path())
+	}
+
+	{
+		xpad := p.X.Padding
+		ypad := p.Y.Padding
+		defer func() {
+			p.X.Padding = xpad
+			p.Y.Padding = ypad
+		}()
+	}
+
+	p.X.Padding = 0
```

Comment:
I am open to suggestions :)
I actually tried to come up with a new interface to encompass this padding issue... to no avail.

as you implied, one way to fix this is to shift things around:

- remove Axis.Padding
- add XAxisPadding and YAxisPadding fields inside each of the various `XyzPlotStyle` structs

possibly do the same for all the other style-dependent fields of `plot.Plot`, `plot.Axis` and `plot.Legend` ?
_ie:_ completely separate "data" from "style" for `plot.Plot`, `plot.Axis` and `plot.Legend`.
then, `plot.Plot` would look like:

```go
type Plot struct {
    Title string
     X, Y Axis
     Legend Legend
     Style draw.PlotStyle
    plotters []Plotter
}

type Axis struct {
   Min, Max float64
   Label string
   Scale Normalizer
}
```

and define, say, in `vg/draw`:

```go
type PlotStyle {
   Title {
     Padding vg.Length
     draw.TextStyle
   }
   BackgroundColor color.Color
   X, Y AxisStyle
   Legend LegendStyle
}

type AxisStyle struct {
   Label TextStyle
   LineStyle
   Padding vg.Length
}

type LegendStyle struct {
   draw.TextStyle
   Padding vg.Length
   Top, Left bool
}
```

having said that...

to flip it around: imagine the default style of `plot.Plot` was the gnuplot-like one, _ie_ with 4 borders delimiting the canvas, and thus with 4 axes to draw.
then, implementing the current `plot.Plot` style would consist in ignoring 2 of the axes delimiting the canvas. :)

another example: nominal axes. the current code for NominalX actually overrides whatever was inside `Axis.Padding`: https://github.com/gonum/plot/blob/master/plot.go#L396

actually: another possibility is to modify `plot.New() (*Plot, error)` to take an optional `plot.Style` which would have a default `MakeAxis` field function (or method) to be used in place of `plot.makeAxis`. (so the possible inconsistency of `plot.Plot.{X,Y}.Padding` would be solved by default)

so... which option would you be more in favor of?